File: hoh/hoh/report/abrechnung_vertriebsmitarbeiter/abrechnung_vertriebsmitarbeiter.py
# Copyright (c) 2021, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils.data import getdate

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": _(""), "fieldname": "platzhalter", "fieldtype": "Data", "width": 150},
        {"label": _("Net Amount"), "fieldname": "net_amount", "fieldtype": "Currency", "width": 120}
    ]

def get_data(filters):
    data = []
    from_date = getdate(filters.from_date)
    to_date = getdate(filters.to_date)
    # get net revenue
    sum_net_base_total = frappe.db.sql("""SELECT 
        SUM(`tabSales Invoice`.`base_net_total`) AS `base_net_total`
    FROM `tabSales Invoice` 
    LEFT JOIN `tabSales Team` ON `tabSales Invoice`.`name` = `tabSales Team`.`parent`
    WHERE `tabSales Team`.`sales_person` = '{sales_person}' AND `tabSales Invoice`.`docstatus` = '1'
      AND `tabSales Invoice`.`posting_date` BETWEEN '{from_date}' AND '{to_date}'
    """.format(sales_person=filters.sales_person, from_date=from_date, to_date=to_date), as_dict=True)
    net_revenue = sum_net_base_total[0]['base_net_total'] or 0
    data.append({
        'platzhalter': _("Nettoumsatz Total"),
        'net_amount': net_revenue
    })

    # get transport deduction
    deduction = frappe.db.sql("""SELECT SUM(`debit`) AS `debit` 
      FROM `tabGL Entry` 
      WHERE `account` = '7300 - Transporte durch Dritte - HOH' 
        AND `tabGL Entry`.`posting_date` BETWEEN '{from_date}' AND '{to_date}'
    """.format(from_date=from_date, to_date=to_date), as_dict=True)
    deductions = (-1) * (deduction[0]['debit'] or 0)
    data.append({
        'platzhalter': _("Transportkosten Total"),
        'net_amount': deductions
    })

    # get transport deduction
    deduction2 = frappe.db.sql("""SELECT SUM(`debit`) AS `debit` 
      FROM `tabGL Entry` 
      WHERE `account` LIKE '7790%' 
        AND `tabGL Entry`.`posting_date` BETWEEN '{from_date}' AND '{to_date}'
    """.format(from_date=from_date, to_date=to_date), as_dict=True)
    deductions2 = (-1) * (deduction2[0]['debit'] or 0)
    data.append({
        'platzhalter': _("Bankgebühren"),
        'net_amount': deductions2
    })
    
    # intermediate sum
    intermediate = net_revenue + deductions + deductions2
    data.append({
        'platzhalter': _("Zwischensumme"),
        'net_amount': intermediate
    })

    # commission
    commission = 0.03 * intermediate
    data.append({
        'platzhalter': _("Kommission"),
        'net_amount': commission
    })
    
    return data

File: hoh/hoh/report/abrechnung_vertriebsmitarbeiter/abrechnung_vertriebsmitarbeiter.js
// Copyright (c) 2021, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Abrechnung Vertriebsmitarbeiter"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From date"),
            "fieldtype": "Date",
            "default": get_start_last_quarter(),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To date"),
            "fieldtype": "Date",
            "default": get_end_last_quarter(),
            "reqd": 1
        },
        {
            "fieldname": "sales_person",
            "label": __("Sales Person"),
            "fieldtype": "Link",
            "options": "Sales Person",
            "reqd": 1
        }
    ]
};

function get_start_last_quarter() {
    var today = new Date(frappe.datetime.get_today());
    var act_month = today.getMonth();
    var fullyear = today.getFullYear();
    if ([0, 1, 2].includes(act_month)) {
        fullyear = fullyear - 1;
        var start = new Date(String(fullyear) + "-10-01");
    }
    if ([3, 4, 5].includes(act_month)) {
        var start = new Date(String(fullyear) + "-01-01");
    }
    if ([6, 7, 8].includes(act_month)) {
        var start = new Date(String(fullyear) + "-04-01");
    }
    if ([9, 10, 11].includes(act_month)) {
        var start = new Date(String(fullyear) + "-07-01");
    }
    return start;
}

function get_end_last_quarter() {
    var today = new Date(frappe.datetime.get_today());
    var act_month = today.getMonth();
    var fullyear = today.getFullYear();
    if ([0, 1, 2].includes(act_month)) {
        fullyear = fullyear - 1;
        var end = new Date(String(fullyear) + "-12-31");
    }
    if ([3, 4, 5].includes(act_month)) {
        var end = new Date(String(fullyear) + "-03-31");
    }
    if ([6, 7, 8].includes(act_month)) {
        var end = new Date(String(fullyear) + "-06-30");
    }
    if ([9, 10, 11].includes(act_month)) {
        var end = new Date(String(fullyear) + "-09-30");
    }
    
    return end;
}
